engine/src/test/java/io/camunda/zeebe/engine/processing/usertask/AssignUserTaskTest.java

```diff
@@ -0,0 +1,190 @@
+    return process(b -> {});
+  }
+
+  private static BpmnModelInstance process(final Consumer<UserTaskBuilder> consumer) {
+    final var builder =
+        Bpmn.createExecutableProcess(PROCESS_ID).startEvent().userTask("task").zeebeUserTask();
+
+    consumer.accept(builder);
+
+    return builder.endEvent().done();
+  }
+
+  @Test
+  public void shouldEmitAssigningEventForUserTask() {
+    // given
+    ENGINE.deployment().withXmlResource(process()).deploy();
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+    final long userTaskKey =
+        RecordingExporter.userTaskRecords(UserTaskIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .getFirst()
+            .getKey();
+
+    // when
+    final Record<UserTaskRecordValue> assignedRecord =
+        ENGINE.userTask().withKey(userTaskKey).withAssignee("foo").assign();
+
+    // then
+    final UserTaskRecordValue recordValue = assignedRecord.getValue();
+
+    Assertions.assertThat(assignedRecord)
+        .hasRecordType(RecordType.EVENT)
+        .hasIntent(UserTaskIntent.ASSIGNING);
+
+    Assertions.assertThat(recordValue)
+        .hasUserTaskKey(userTaskKey)
+        .hasTenantId(TenantOwned.DEFAULT_TENANT_IDENTIFIER);
+  }
+
+  @Test
+  public void shouldRejectAssignIfUserTaskNotFound() {
+    // given
+    final int key = 123;
+
+    // when
+    final Record<UserTaskRecordValue> assignedRecord =
+        ENGINE.userTask().withKey(key).withAssignee("foo").expectRejection().assign();
+
+    // then
+    Assertions.assertThat(assignedRecord).hasRejectionType(RejectionType.NOT_FOUND);
+  }
+
+  @Test
+  public void shouldAssignUserTaskWithAssignee() {
+    // given
+    ENGINE.deployment().withXmlResource(process()).deploy();
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    // when
+    final Record<UserTaskRecordValue> assignedRecord =
+        ENGINE.userTask().ofInstance(processInstanceKey).withAssignee("foo").assign();
+
+    // then
+    Assertions.assertThat(assignedRecord)
+        .hasRecordType(RecordType.EVENT)
+        .hasIntent(UserTaskIntent.ASSIGNING);
+    Assertions.assertThat(assignedRecord.getValue()).hasAssignee("foo");
+  }
+
+  @Test
+  public void shouldUnassignUserTaskWithNoAssignee() {
+    // given
+    ENGINE.deployment().withXmlResource(process(b -> b.zeebeAssignee("foo"))).deploy();
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    // when
+    final Record<UserTaskRecordValue> assignedRecord =
+        ENGINE.userTask().ofInstance(processInstanceKey).assign();
+
+    // then
+    Assertions.assertThat(assignedRecord)
+        .hasRecordType(RecordType.EVENT)
+        .hasIntent(UserTaskIntent.ASSIGNING);
+    assertThat(assignedRecord.getValue().getAssignee()).isEmpty();
+  }
+
+  @Test
+  public void shouldRejectAssignIfUserTaskIsCompleted() {
```

Comment:
:wrench: It would have been nice if we had a test case to reject an ASSIGN command if the user task is already in the `ASSIGNING` state, but it's very difficult to do that now.

:question: Maybe we can create an issue for these types of test cases, so we don't forget to add them once these scenarios become possible to test?

Reply:
I agree, I was also thinking "How can I get in between there now to test this?" 😄
Will be possible with user task listeners then. I currently don't see any other way that's not completely artificial and too complex. But maybe you know of a way to have the engine not execute the ASSIGNED applier for the test? 😄

👍 I added this to the Task Listener PDP epic for now: https://github.com/camunda/product-hub/issues/1822#issuecomment-1854085499